File: interviews/ForUsAll/Python/basicCalculatorII.py
class Solution:
    def calculate(self, s):
        
        if not s:
            return 0 
        
        num = 0 
        sign = '+'
        stack = [] 
        length = len(s)
        
        for i in range(length):
            
            if s[i].isdigit():
                num = num * 10 + int(s[i])
                
            if (s[i].isdigit() == False and s[i] != ' ') or i == length - 1:
                
                if sign == '+':
                    stack.append(num)
                    
                if sign == '-':
                    stack.append(-num)
                    
                if sign == '*':
                    stack.append(stack.pop() * num) 
                    
                if sign == '/':
                    tmp = stack.pop()
                    if tmp // num < 0 and tmp % num != 0:
                        stack.append(tmp // num + 1)
                    else:
                        stack.append(tmp // num)
                    
                sign = s[i]
                num = 0 
                
        res = 0
        
        for n in stack:
            res += n 
            
        return res 
        

File: interviews/ForUsAll/Java/BasicCalculatorII.java
import java.util.*;

public class BasicCalculatorII {
    public int calculate(String s) {
        
        if (s == null || s.length() == 0) {
            return 0;
        }
        
        int len = s.length();
        
        char sign = '+';
        int num = 0; 
        
        Stack<Integer> stack = new Stack<>(); 
        
        for (int i = 0; i < len; i++) {
            
            if (Character.isDigit(s.charAt(i))) num = num * 10 + s.charAt(i) - '0';// digit
            
            if ( (!Character.isDigit(s.charAt(i)) && s.charAt(i) != ' ') || i == len - 1) {
                
                if (sign == '-') stack.push(-num);
                
                if (sign == '+') stack.push(num);
                
                if (sign == '*') stack.push(stack.pop() * num);
                
                if (sign == '/') stack.push(stack.pop() / num);
                
                sign = s.charAt(i);
                num = 0;
            }
            
        }
        
        int re = 0;
        
        for (Integer n : stack) {
            re += n;
        }
        
        return re;
    }
}